### PR TITLE
Make NaNs quiet by default and other NaN tidy-up

### PR DESCRIPTION
**Pull request by LewisGaul**
Reviving https://github.com/ziglang/zig/pull/12361 (previously https://github.com/ziglang/zig/pull/10432). Every time I reraise the PR it gets better, so maybe this should be closed ;)

- Make `std.math.nan()` return canonical quiet NaN, rather than signalling NaN which is not properly supported yet
  - This is needed by https://github.com/ziglang/zig/pull/16808
- Fix up some basic support for signalling NaN:
  - Implement `std.math.snan()` to return a somewhat canonical signalling NaN value (better than nothing, even if it's not *the* canonical value on all platforms), checked f80 at https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format
  - Implement `std.math.isSignalNan()` properly, with testcase
  - Although previously suggested by @tiehuis to remove `std.math.snan()` in [this comment](https://github.com/ziglang/zig/pull/10432#issuecomment-1003666676) and [this comment](https://github.com/ziglang/zig/pull/10432#issuecomment-1003779682), I have opted not to do this to reduce disruption
  - Adjust implementation of behaviour testcases covering signalling NaNs in `test/behavior/bitcast.zig` and `test/behavior/bugs/14198.zig`
- Tidy up, removing `nan.zig` and `std.math.[q]nan_*` constants to handle it the same way as `inf`, following work in:
  - https://github.com/ziglang/zig/pull/10133
  - https://github.com/ziglang/zig/pull/11401
  - https://github.com/ziglang/zig/pull/15957

### CODE REVIEW

File: lib/std/math.zig
Comment (LewisGaul) [inline lib/std/math.zig:107]: These are deliberately all recommended `nan()` (quiet NaN) since there isn't proper support for signalling NaNs yet. Are we happy with these hard deprecations being made immediately, in line with those above?

**Comment by LewisGaul** [comment]
@tiehuis in case you wanted to have a look, having reviewed the initial PR
@matu3ba for interest, relating to https://github.com/ziglang/zig/pull/16808

The tests are all passing, so this should be ready for review/merge :D

**Review by matu3ba — CHANGES_REQUESTED** [review]
I like the change, but took only a relative quick glimpse. Also good that CI works. :) I do phrase it here as "miscompilations", but it could be also be seen as "relaxation of float semantics".

1. Please make all known miscompilations explicit to user or at least hint what we know and please test all known not miscompilations (so far).

2. Also, please add a comment to the test that the "known not miscompilations" provide no guarantee, that llvm will not miscompile other code.

3. I am not sure, what Zig should or could provide "to test for known possible miscompilations by LLVM" (signaling nan get lost and become silent nan). However, this can also be clarified before or after merge by the core member.

File: lib/std/math/isnan.zig
Comment (andrewrk) [inline lib/std/math/isnan.zig:6]: This documentation is redundant with the function name.

File: lib/std/math/isnan.zig
Comment (andrewrk) [inline lib/std/math/isnan.zig:11]: This documentation is redundant with the function name.

**Review by andrewrk — APPROVED** [review]
Looks ready to merge to me - I'll give @tiehuis a few days to review in case he wants to, otherwise will proceed with the merge.

File: lib/std/math/float.zig
Comment (LewisGaul) [reply lib/std/math/float.zig:111]: I've added a note to the docstring pointing to https://github.com/ziglang/zig/issues/14366. I'm reluctant to try and make more specific statements given I'm not sure exactly what's true, I think it needs more investigation, and this PR doesn't actually make things any *worse* - I've pretty much just moved this function and tidied up the implementation.

File: lib/std/math/isnan.zig
Comment (LewisGaul) [reply lib/std/math/isnan.zig]: Would this be better placed in `test/behavior/bitcast.zig` (alongside "bitcast nan float does not modify signaling bit") or a new `test/behavior/14366.zig` file? Again, I'm tempted to consider this out of scope - I'm not making things any *worse* in this PR, just tidying up some implementations. I'd be happy to keep looking into this stuff after this PR goes in though.

File: test/behavior/bitcast.zig
Comment (LewisGaul) [inline test/behavior/bitcast.zig:416]: This testcase is now pretty much a duplicate of the "math.snan" test I added in `lib/std/math/float.zig`, and there's also duplication in `test/behavior/bugs/14198.zig` - any suggestions for [re]moving either of these behaviour tests? Perhaps adding a new `test/behavior/14366.zig` file corresponding to https://github.com/ziglang/zig/issues/14366 (as suggested in https://github.com/ziglang/zig/pull/16826#discussion_r1297753936)?

File: lib/std/math/isnan.zig
Comment (matu3ba) [reply lib/std/math/isnan.zig]: >  Again, I'm tempted to consider this out of scope - I'm not making things any worse in this PR, just tidying up some implementations. I'd be happy to keep looking into this stuff after this PR goes in though.

I'm happy with either decision from your side. There is a TODO and the context is on github and in the PR etc. Let's see what @tiehuis think.

Comment (tiehuis) [reply lib/std/math/isnan.zig:33]: I agree. Can leave this as-is and move later if deemed important.

File: test/behavior/bitcast.zig
Comment (tiehuis) [reply test/behavior/bitcast.zig:416]: I think what exists currently is fine even with the duplication. Even if they are inherently testing the same thing would consider them separate logically and makes sense to test at both sites. Also not a lot of code.

Would be happy to remove `test/behaviour/bugs/14198.zig` and just keeping this behaviour test.

**Comment by tiehuis** [comment]
Looks good to me. One possible action here https://github.com/ziglang/zig/pull/16826#discussion_r1298025778 but wouldn't block this PR and can grab in a follow-up.

**Review by tiehuis — APPROVED** [review]

**Comment by LewisGaul** [comment]
Thanks all for taking a look, very excited to have the first step of my maths work merged after more than a year and a half! (the elapsed time being no fault of anyone else)

**Pull request merged**
